File: src/code_indexing_mcp/installer/update.py
"""Self-update for a managed installation, in two phases.

Phase one runs from the environment it is about to rewrite: it fast-forwards
the checkout, re-syncs the locked environment, and then hands off to a child
process that starts from the new files on disk. Phase two (``--finalize``)
reconciles the accelerator environment, stops the daemon, re-runs the
post-install checks, and reports what moved.

Every import in this module is at the top for that reason: the module is
imported before the pull, and once the files on disk have changed an import
resolved later would load new code into a process running the old code.
"""

from __future__ import annotations

import contextlib
import json
import os
import shutil
import subprocess
import sys
import time
from collections.abc import Callable
from pathlib import Path
from typing import NamedTuple

from filelock import FileLock, Timeout

from .. import update_check
from ..application import RuntimePaths
from ..daemon import BrokerApplication, daemon_status, daemon_supported
from .accelerator import (
    ACCELERATOR_EXTRAS,
    _run_command,
    _uv_executable,
    accelerator_lock_fingerprint,
    accelerator_record_path,
    configure_accelerator,
    environment_python,
    prepared_accelerator,
    server_executable,
)
from .config_files import InstallerError
from .harnesses import configuration_path
from .orchestrator import default_install_directory
from .verify import format_check, run_update_checks
from .wizard import load_prefill

# Kept in lockstep with install.py, which must stay self-contained: it is
# downloaded and run on its own, before this package exists on the machine.
DEFAULT_REPOSITORY_URL = "https://github.com/MarcinHamiga/code-indexing-mcp.git"
SERVING_EXTRAS = ("cpu", "tui")

UPDATE_BRANCH = "main"
REMOTE_BRANCH_REF = "refs/heads/main"
UPDATE_LOCK_NAME = ".update.lock"
# Suffix for a Windows launcher renamed out of the way of `uv sync`.
STALE_SUFFIX = ".stale-"
LOG_LINE_LIMIT = 15
DIRTY_PATH_LIMIT = 5
LS_REMOTE_TIMEOUT_SECONDS = 10.0
# Distinct from 1 (error) and 2 (CodeIndexingError), so a script can tell
# "an update is available" from "the check itself failed".
CHECK_UPDATE_AVAILABLE_EXIT = 10

RunCommand = Callable[..., "subprocess.CompletedProcess[str]"]
Spawn = Callable[[list[str], Path], int]


class _AcceleratorOutcome(NamedTuple):
    status: str  # "ok" | "skipped" | "warning"
    detail: str
    rebuilt: bool
    prepared: str | None


def _canonical_repository_url(url: str) -> str:
    value = url.strip().rstrip("/")
    if value.endswith(".git"):
        value = value[:-4]
    if value.startswith("git@github.com:"):
        return f"github.com/{value.removeprefix('git@github.com:').lower()}"
    for prefix in ("https://github.com/", "http://github.com/", "ssh://git@github.com/"):
        if value.startswith(prefix):
            return f"github.com/{value.removeprefix(prefix).lower()}"
    if "://" not in value:
        return str(Path(value).expanduser().resolve())
    return value


def _error(message: str) -> int:
    print(f"Error: {message}", file=sys.stderr)
    return 1


def _print_status(step: str, status: str, detail: str) -> None:
    stream = sys.stderr if status == "warning" else sys.stdout
    print(f"[{step}] {status}: {detail}", file=stream)


def _git(run_command: RunCommand, directory: Path, *arguments: str) -> str:
    return run_command(["git", *arguments], cwd=directory).stdout.strip()


def _fast_forward_possible(directory: Path) -> bool:
    """Ask git whether HEAD is an ancestor of what was just fetched.

    This one call answers through its exit status, so it cannot go through the
    command helper every other call uses -- that one raises on non-zero.
    """

    try:
        completed = subprocess.run(
            ["git", "merge-base", "--is-ancestor", "HEAD", "FETCH_HEAD"],
            cwd=directory,
            capture_output=True,
            text=True,
            check=False,
        )
    except OSError:
        return False
    return completed.returncode == 0


def _spawn_finalize(argv: list[str], cwd: Path) -> int:
    # subprocess rather than os.execv: exec is unavailable on Windows in any
    # form that survives a running console, and it would forfeit the chance to
    # report a launch failure at all.
    try:
        return subprocess.run(argv, cwd=cwd).returncode
    except OSError as exc:
        return _error(f"could not start the updated code to finish the update: {exc}")


def update_main(
    *,
    install_dir: str | None,
    check: bool,
    skip_accelerator: bool,
    finalize: bool,
    previous_sha: str | None,
    run_command: RunCommand = _run_command,
    uv_executable: str | None = None,
    platform_name: str | None = None,
    spawn: Spawn = _spawn_finalize,
) -> int:
    """Entry for ``code-indexing-mcp update`` and its two hidden sub-modes."""

    directory = (
        Path(install_dir).expanduser().resolve()
        if install_dir
        else default_install_directory().resolve()
    )
    if finalize:
        return _finalize_main(
            directory,
            previous_sha=previous_sha,
            skip_accelerator=skip_accelerator,
            run_command=run_command,
        )
    if check:
        return _check_main(directory, run_command=run_command)
    return _phase1(
        directory,
        skip_accelerator=skip_accelerator,
        run_command=run_command,
        uv_executable=uv_executable,
        platform_name=platform_name,
        spawn=spawn,
    )


def _preflight(
    directory: Path,
    *,
    run_command: RunCommand,
    uv_executable: str | None,
) -> tuple[str, str, str]:
    """Return ``(uv, head, remote)``, or raise before anything has been touched."""

    if shutil.which("git") is None:
        raise InstallerError("git is required to update but was not found in PATH")
    if not (directory / ".git").exists():
        raise InstallerError(
            f"{directory} is not a git checkout; reinstall with install.sh to update it "
            "with this command"
        )
    try:
        origin = _git(run_command, directory, "remote", "get-url", "origin")
    except InstallerError as exc:
        raise InstallerError(f"{directory} has no origin remote to update from: {exc}") from exc
    expected = os.environ.get("CODE_INDEXING_MCP_REPO_URL", DEFAULT_REPOSITORY_URL)
    if _canonical_repository_url(origin) != _canonical_repository_url(expected):
        raise InstallerError(
            f"the checkout at {directory} tracks {origin}, not {expected}; update it manually "
            "or point CODE_INDEXING_MCP_REPO_URL at the remote it came from"
        )
    branch = _git(run_command, directory, "rev-parse", "--abbrev-ref", "HEAD")
    if branch != UPDATE_BRANCH:
        raise InstallerError(
            f"the checkout at {directory} is on branch {branch}, not {UPDATE_BRANCH}; "
            f"switch it back with `git switch {UPDATE_BRANCH}` before updating"
        )
    # Tracked changes only: the environment, its caches, and this command's own
    # lock file all live untracked inside the checkout, and none of them is a
    # reason to refuse a fast-forward.
    status = _git(run_command, directory, "status", "--porcelain", "--untracked-files=no")
    dirty = status.splitlines()
    if dirty:
        names = ", ".join(line.strip().split(maxsplit=1)[-1] for line in dirty[:DIRTY_PATH_LIMIT])
        if len(dirty) > DIRTY_PATH_LIMIT:
            names = f"{names} and {len(dirty) - DIRTY_PATH_LIMIT} more"
        raise InstallerError(
            f"the checkout at {directory} has uncommitted changes ({names}); commit, stash, "
            "or discard them before updating"
        )
    # Resolved before the pull on purpose: a checkout left ahead of an
    # environment that cannot be synced is the one state this refuses to create.
    uv = _uv_executable(uv_executable)
    try:
        run_command(["git", "fetch", "origin", UPDATE_BRANCH], cwd=directory)
    except InstallerError as exc:
        raise InstallerError(
            f"could not fetch origin/{UPDATE_BRANCH}; check the network connection and try "
            f"again: {exc}"
        ) from exc
    head = _git(run_command, directory, "rev-parse", "HEAD")
    remote = _git(run_command, directory, "rev-parse", "FETCH_HEAD")
    if head != remote and not _fast_forward_possible(directory):
        raise InstallerError(
            f"the checkout at {directory} has diverged from origin/{UPDATE_BRANCH}; update it "
            "manually or reinstall"
        )
    return uv, head, remote


def _phase1(
    directory: Path,
    *,
    skip_accelerator: bool,
    run_command: RunCommand,
    uv_executable: str | None,
    platform_name: str | None,
    spawn: Spawn,
) -> int:
    if not server_executable(directory, platform_name=platform_name).is_file():
        return _error(f"no installation found at {directory}")
    lock = FileLock(directory / UPDATE_LOCK_NAME)
    try:
        lock.acquire(timeout=0)
    except Timeout:
        return _error(f"another update is already running in {directory}; wait for it to finish")
    try:
        _discard_stale_scripts(directory, platform_name=platform_name)
        try:
            uv, head, remote = _preflight(
                directory, run_command=run_command, uv_executable=uv_executable
            )
        except InstallerError as exc:
            return _error(str(exc))
        if head == remote:
            # Still synced and finalized below: an interrupted update leaves the
            # checkout current and everything after it undone.
            print("Already up to date.")
        else:
            try:
                run_command(["git", "merge", "--ff-only", "FETCH_HEAD"], cwd=directory)
            except InstallerError as exc:
                return _error(f"could not fast-forward the checkout at {directory}: {exc}")
        shielded = _shield_running_script(directory, platform_name=platform_name)
        extras = [flag for extra in SERVING_EXTRAS for flag in ("--extra", extra)]
        try:
            run_command([uv, "sync", "--locked", *extras], cwd=directory)
        except InstallerError as exc:
            _unshield(shielded, directory, platform_name=platform_name)
            return _error(
                f"{exc}\nthe checkout was updated but its environment was not; re-run "
                "`code-indexing-mcp update`"
            )
        _unshield(shielded, directory, platform_name=platform_name)
        python = environment_python(directory / ".venv", platform_name=platform_name)
        argv = [
            str(python),
            "-m",
            "code_indexing_mcp",
            "update",
            "--finalize",
            "--previous-sha",
            head,
        ]
        if skip_accelerator:
            argv.append("--skip-accelerator")
        argv += ["--install-dir", str(directory)]
        return spawn(argv, directory)
    finally:
        lock.release()


def _shield_running_script(directory: Path, *, platform_name: str | None = None) -> Path | None:
    """Rename the running launcher aside so ``uv sync`` can replace it.

    Windows refuses to overwrite a mapped image but allows renaming one, which
    is what makes an update that runs through this very launcher possible.
    """

    platform_name = platform_name or sys.platform
    if not platform_name.startswith("win"):
        return None
    executable = server_executable(directory, platform_name=platform_name)
    if not executable.is_file():
        return None
    stale = executable.with_name(f"{executable.name}{STALE_SUFFIX}{os.getpid()}")
    try:
        executable.rename(stale)
    except OSError:
        return None
    return stale


def _unshield(shielded: Path | None, directory: Path, *, platform_name: str | None = None) -> None:
    """Put the launcher back, unless uv wrote a new one over the gap."""

    if shielded is None:
        return
    executable = server_executable(directory, platform_name=platform_name)
    if not executable.exists():
        with contextlib.suppress(OSError):
            shielded.rename(executable)
        return
    # An unlink that fails leaves a copy this process still has mapped; the
    # next update collects it.
    with contextlib.suppress(OSError):
        shielded.unlink()


def _discard_stale_scripts(directory: Path, *, platform_name: str | None = None) -> None:
    platform_name = platform_name or sys.platform
    if not platform_name.startswith("win"):
        return
    scripts = directory / ".venv" / "Scripts"
    try:
        leftovers = list(scripts.glob(f"*{STALE_SUFFIX}*"))
    except OSError:
        return
    for path in leftovers:
        try:
            path.unlink()
        except OSError:
            continue


def _recorded_lock_fingerprint(directory: Path) -> str:
    try:
        record = json.loads(accelerator_record_path(directory).read_text(encoding="utf-8"))
    except (OSError, ValueError, InstallerError):
        return ""
    if not isinstance(record, dict):
        return ""
    return str(record.get("lock_fingerprint", ""))


def _reconcile_accelerator(directory: Path, *, skip_accelerator: bool) -> _AcceleratorOutcome:
    """Rebuild the accelerator environment only when the pull invalidated it."""

    prepared = prepared_accelerator(directory)
    if prepared is None:
        # An update never detects: an installation that is on CPU stays on CPU
        # until someone asks for something else.
        return _AcceleratorOutcome("skipped", "no accelerator environment is recorded", False, None)
    extra = ACCELERATOR_EXTRAS.get(prepared)
    if extra is None:
        return _AcceleratorOutcome(
            "skipped", f"the recorded {prepared} runtime needs no separate environment", False, None
        )
    try:
        expected = accelerator_lock_fingerprint(directory, extra)
    except InstallerError as exc:
        return _AcceleratorOutcome(
            "warning", f"the {prepared} environment could not be checked: {exc}", False, prepared
        )
    if _recorded_lock_fingerprint(directory) == expected:
        # Matching fingerprints answer the question without a single detection
        # subprocess, the slowest of which can sit on nvidia-smi for ~30s.
        return _AcceleratorOutcome(
            "skipped",
            f"the {prepared} environment is unchanged since the last build",
            False,
            prepared,
        )
    if skip_accelerator:
        return _AcceleratorOutcome(
            "warning",
            f"the {prepared} environment was resolved from an older lockfile and keeps serving "
            f"as it is, because the server does not re-check it; rebuild it with "
            f"`code-indexing-mcp configure --accelerator {prepared}`",
            False,
            prepared,
        )
    try:
        plan = configure_accelerator(directory, prepared)
    except InstallerError as exc:
        return _AcceleratorOutcome(
            "warning", f"the {prepared} environment could not be rebuilt: {exc}", False, prepared
        )
    # A rebuild that fell back has already cleared the record and degraded to
    # CPU, which costs speed and nothing else -- a warning, not a failed update.
    return _AcceleratorOutcome(
        "ok" if plan.honored else "warning",
        f"{plan.accelerator} ({plan.reason})",
        True,
        plan.accelerator if plan.prepares_environment else None,
    )


def _wait_until_stopped(
    paths: RuntimePaths, *, attempts: int = 100, interval: float = 0.05
) -> bool:
    for _ in range(attempts):
        if not daemon_status(paths)["running"]:
            return True
        time.sleep(interval)
    return False


def _stop_daemon(paths: RuntimePaths, *, changed: bool) -> tuple[str, str]:
    """Stop the daemon; the next client respawns it on the updated code."""

    if not changed:
        return "skipped", "nothing changed, so the running daemon is already current"
    if not daemon_supported():
        return "skipped", "this platform has no shared daemon"
    try:
        if not daemon_status(paths)["running"]:
            return "skipped", "no daemon is running"
        BrokerApplication(paths).stop()
        if not _wait_until_stopped(paths):
            return "warning", "the daemon did not stop; run `code-indexing-mcp daemon stop`"
    except Exception as exc:
        return "warning", f"the daemon could not be stopped: {exc}"
    return "ok", "stopped; it restarts on the updated code with the next client"


def _print_summary(
    directory: Path,
    *,
    previous_sha: str | None,
    head: str,
    accelerator: _AcceleratorOutcome,
    run_command: RunCommand,
) -> None:
    if previous_sha and previous_sha != head:
        print(f"Updated {previous_sha[:7]} -> {head[:7]}")
        try:
            lines = _git(
                run_command,
                directory,
                "log",
                "--oneline",
                "--no-decorate",
                f"{previous_sha}..HEAD",
            ).splitlines()
        except InstallerError:
            lines = []
        for line in lines[:LOG_LINE_LIMIT]:
            print(f"  {line}")
        if len(lines) > LOG_LINE_LIMIT:
            print(f"  ... and {len(lines) - LOG_LINE_LIMIT} more commits")
    else:
        print(f"Already at {head[:7]}; the environment was re-synced and re-checked.")
    print(f"Accelerator: {accelerator.detail}")
    print("Restart your MCP clients to load the updated server.")


def _finalize_main(
    directory: Path,
    *,
    previous_sha: str | None,
    skip_accelerator: bool,
    run_command: RunCommand,
) -> int:
    head = update_check.checkout_head(directory) or ""
    accelerator = _reconcile_accelerator(directory, skip_accelerator=skip_accelerator)
    _print_status("accelerator", accelerator.status, accelerator.detail)

    paths = RuntimePaths.from_environment()
    changed = accelerator.rebuilt or (previous_sha or "") != head
    status, detail = _stop_daemon(paths, changed=changed)
    _print_status("daemon", status, detail)

    configured = [(slug, configuration_path(slug)) for slug in load_prefill().configured_slugs]
    checks = run_update_checks(
        directory, configured, accelerator_was_prepared=accelerator.prepared is not None
    )
    for check in checks:
        print(format_check(check), file=sys.stdout if check.ok else sys.stderr)

    # Silencing the notifier is a courtesy; failing to is not a failed update.
    with contextlib.suppress(OSError):
        update_check.write_cache(
            paths.cache,
            update_check.UpdateStatus(checked_at=time.time(), local_sha=head, remote_sha=head),
        )

    _print_summary(
        directory,
        previous_sha=previous_sha,
        head=head,
        accelerator=accelerator,
        run_command=run_command,
    )
    unusable = [
        check for check in checks if check.status == "fail" and check.name == "server executable"
    ]
    return 1 if unusable else 0


def _check_main(directory: Path, *, run_command: RunCommand) -> int:
    if not (directory / ".git").exists():
        return _error(f"{directory} is not a git checkout; there is no update to check for")
    try:
        origin = _git(run_command, directory, "remote", "get-url", "origin")
    except InstallerError as exc:
        return _error(f"{directory} has no origin remote to check against: {exc}")
    expected = os.environ.get("CODE_INDEXING_MCP_REPO_URL", DEFAULT_REPOSITORY_URL)
    if _canonical_repository_url(origin) != _canonical_repository_url(expected):
        return _error(f"the checkout at {directory} tracks {origin}, not {expected}")
    try:
        completed = subprocess.run(
            ["git", "ls-remote", "origin", REMOTE_BRANCH_REF],
            cwd=directory,
            capture_output=True,
            text=True,
            timeout=LS_REMOTE_TIMEOUT_SECONDS,
            check=True,
        )
    except (OSError, subprocess.SubprocessError) as exc:
        return _error(f"could not reach origin to check for updates: {exc}")
    lines = [line for line in completed.stdout.splitlines() if line.strip()]
    if not lines:
        return _error(f"origin has no {REMOTE_BRANCH_REF} to compare against")
    status = update_check.UpdateStatus(
        checked_at=time.time(),
        local_sha=update_check.checkout_head(directory) or "",
        remote_sha=lines[0].split()[0],
    )
    print(
        json.dumps(
            {
                "install_dir": str(directory),
                "local_sha": status.local_sha,
                "remote_sha": status.remote_sha,
                "update_available": status.update_available,
            },
            indent=2,
            sort_keys=True,
        )
    )
    with contextlib.suppress(OSError):
        update_check.write_cache(RuntimePaths.from_environment().cache, status)
    return CHECK_UPDATE_AVAILABLE_EXIT if status.update_available else 0

File: ts/packages/installer/src/update.ts
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import {
  ACCELERATOR_PROBE_TARGETS,
  acceleratorLockFingerprint,
  acceleratorRecordPath,
  configureAccelerator,
  preparedAccelerator,
  serverExecutable,
  writeServerLauncher,
} from "./accelerator.ts";
import { canonicalRepositoryUrl, DEFAULT_REPOSITORY_URL } from "./bootstrap.ts";
import { InstallerError, resolveExisting } from "./config-files.ts";
import { commandFromEntry } from "./env-blocks.ts";
import { configurationPath, installSkills, readServerEntry, skillDirectory } from "./harnesses.ts";
import { defaultInstallDirectory } from "./orchestrator.ts";
import { formatCheck, runUpdateChecks } from "./verify.ts";
import { loadPrefill } from "./wizard.ts";

export const UPDATE_BRANCH = "main";
export const UPDATE_LOCK_NAME = ".update.lock";
export const CHECK_UPDATE_AVAILABLE_EXIT = 10;
const LOG_LINE_LIMIT = 15;
const DIRTY_PATH_LIMIT = 5;
const LS_REMOTE_TIMEOUT_SECONDS = 10;

export type RunCommand = (
  arguments_: string[],
  options?: { cwd?: string },
) => { stdout: string; status: number };
export type Spawn = (argv: string[], cwd: string) => number;

function defaultRunCommand(
  arguments_: string[],
  options: { cwd?: string } = {},
): { stdout: string; status: number } {
  const result = spawnSync(arguments_[0] ?? "", arguments_.slice(1), {
    cwd: options.cwd,
    encoding: "utf8",
  });
  if (result.error !== undefined) {
    throw new InstallerError(`Required command was not found: ${arguments_[0]}`);
  }
  if (result.status !== 0) {
    const detail = (result.stderr || result.stdout || "").trim();
    throw new InstallerError(
      detail === ""
        ? `Command failed: ${arguments_.join(" ")}`
        : `Command failed: ${arguments_.join(" ")}\n${detail}`,
    );
  }
  return { stdout: result.stdout, status: result.status ?? 0 };
}

function defaultSpawn(argv: string[], cwd: string): number {
  const result = spawnSync(argv[0] ?? "", argv.slice(1), { cwd, stdio: "inherit" });
  if (result.error !== undefined) {
    process.stderr.write(
      `Error: could not start the updated code to finish the update: ${result.error.message}\n`,
    );
    return 1;
  }
  return result.status ?? 1;
}

function error(message: string): number {
  process.stderr.write(`Error: ${message}\n`);
  return 1;
}

function printStatus(step: string, status: string, detail: string): void {
  const stream = status === "warning" ? process.stderr : process.stdout;
  stream.write(`[${step}] ${status}: ${detail}\n`);
}

function git(runCommand: RunCommand, directory: string, ...arguments_: string[]): string {
  return runCommand(["git", ...arguments_], { cwd: directory }).stdout.trim();
}

function fastForwardPossible(directory: string): boolean {
  const result = spawnSync("git", ["merge-base", "--is-ancestor", "HEAD", "FETCH_HEAD"], {
    cwd: directory,
    encoding: "utf8",
  });
  return result.status === 0;
}

function acquireLock(directory: string): () => void {
  const lockPath = path.join(directory, UPDATE_LOCK_NAME);
  try {
    const handle = fs.openSync(lockPath, "wx");
    fs.writeFileSync(handle, String(process.pid));
    fs.closeSync(handle);
  } catch {
    throw new InstallerError(
      `another update is already running in ${directory}; wait for it to finish`,
    );
  }
  return () => {
    fs.rmSync(lockPath, { force: true });
  };
}

export async function updateMain(options: {
  installDir?: string | null;
  check?: boolean;
  skipAccelerator?: boolean;
  finalize?: boolean;
  previousSha?: string | null;
  runCommand?: RunCommand;
  spawn?: Spawn;
}): Promise<number> {
  const directory =
    options.installDir != null
      ? resolveExisting(options.installDir)
      : resolveExisting(defaultInstallDirectory());
  const runCommand = options.runCommand ?? defaultRunCommand;
  const spawn = options.spawn ?? defaultSpawn;
  if (options.finalize === true) {
    return finalizeMain(directory, {
      previousSha: options.previousSha ?? null,
      skipAccelerator: options.skipAccelerator === true,
      runCommand,
    });
  }
  if (options.check === true) return checkMain(directory, runCommand);
  return phase1(directory, {
    skipAccelerator: options.skipAccelerator === true,
    runCommand,
    spawn,
  });
}

function preflight(directory: string, runCommand: RunCommand): [string, string] {
  if (!fs.existsSync(path.join(directory, ".git"))) {
    throw new InstallerError(
      `${directory} is not a git checkout; reinstall with install.sh to update it with this command`,
    );
  }
  const origin = git(runCommand, directory, "remote", "get-url", "origin");
  const expected = process.env.CODE_INDEXING_MCP_REPO_URL ?? DEFAULT_REPOSITORY_URL;
  if (canonicalRepositoryUrl(origin) !== canonicalRepositoryUrl(expected)) {
    throw new InstallerError(
      `the checkout at ${directory} tracks ${origin}, not ${expected}; update it manually or point CODE_INDEXING_MCP_REPO_URL at the remote it came from`,
    );
  }
  const branch = git(runCommand, directory, "rev-parse", "--abbrev-ref", "HEAD");
  if (branch !== UPDATE_BRANCH) {
    throw new InstallerError(
      `the checkout at ${directory} is on branch ${branch}, not ${UPDATE_BRANCH}; switch it back with \`git switch ${UPDATE_BRANCH}\` before updating`,
    );
  }
  const status = git(runCommand, directory, "status", "--porcelain", "--untracked-files=no");
  const dirty = status === "" ? [] : status.split(/\r?\n/);
  if (dirty.length > 0) {
    const names = dirty
      .slice(0, DIRTY_PATH_LIMIT)
      .map((line) => line.trim().split(/\s+/).at(-1) ?? line)
      .join(", ");
    const extra =
      dirty.length > DIRTY_PATH_LIMIT ? ` and ${dirty.length - DIRTY_PATH_LIMIT} more` : "";
    throw new InstallerError(
      `the checkout at ${directory} has uncommitted changes (${names}${extra}); commit, stash, or discard them before updating`,
    );
  }
  try {
    runCommand(["git", "fetch", "origin", UPDATE_BRANCH], { cwd: directory });
  } catch (cause) {
    throw new InstallerError(
      `could not fetch origin/${UPDATE_BRANCH}; check the network connection and try again: ${cause instanceof Error ? cause.message : String(cause)}`,
    );
  }
  const head = git(runCommand, directory, "rev-parse", "HEAD");
  const remote = git(runCommand, directory, "rev-parse", "FETCH_HEAD");
  if (head !== remote && !fastForwardPossible(directory)) {
    throw new InstallerError(
      `the checkout at ${directory} has diverged from origin/${UPDATE_BRANCH}; update it manually or reinstall`,
    );
  }
  return [head, remote];
}

function phase1(
  directory: string,
  options: { skipAccelerator: boolean; runCommand: RunCommand; spawn: Spawn },
): number {
  if (!fs.existsSync(serverExecutable(directory))) {
    return error(`no installation found at ${directory}`);
  }
  let release: (() => void) | null = null;
  try {
    release = acquireLock(directory);
    let head: string;
    let remote: string;
    try {
      [head, remote] = preflight(directory, options.runCommand);
    } catch (cause) {
      return error(cause instanceof Error ? cause.message : String(cause));
    }
    if (head === remote) process.stdout.write("Already up to date.\n");
    else {
      try {
        options.runCommand(["git", "merge", "--ff-only", "FETCH_HEAD"], { cwd: directory });
      } catch (cause) {
        return error(
          `could not fast-forward the checkout at ${directory}: ${cause instanceof Error ? cause.message : String(cause)}`,
        );
      }
    }
    try {
      options.runCommand([process.execPath, "install", "--frozen-lockfile"], {
        cwd: path.join(directory, "ts"),
      });
    } catch (cause) {
      return error(
        `${cause instanceof Error ? cause.message : String(cause)}\nthe checkout was updated but its environment was not; re-run \`code-indexing-mcp update\``,
      );
    }
    writeServerLauncher(directory);
    const argv = [
      process.execPath,
      path.join(directory, "ts", "packages", "installer", "src", "command.ts"),
      "update",
      "--finalize",
      "--previous-sha",
      head,
      "--install-dir",
      directory,
    ];
    if (options.skipAccelerator) argv.push("--skip-accelerator");
    return options.spawn(argv, directory);
  } catch (cause) {
    return error(cause instanceof Error ? cause.message : String(cause));
  } finally {
    release?.();
  }
}

async function reconcileAccelerator(
  directory: string,
  skipAccelerator: boolean,
): Promise<{ status: string; detail: string; rebuilt: boolean; prepared: string | null }> {
  const prepared = preparedAccelerator(directory);
  if (prepared === null) {
    return {
      status: "skipped",
      detail: "no accelerator environment is recorded",
      rebuilt: false,
      prepared: null,
    };
  }
  if (!ACCELERATOR_PROBE_TARGETS.has(prepared)) {
    return {
      status: "skipped",
      detail: `the recorded ${prepared} runtime needs no separate environment`,
      rebuilt: false,
      prepared: null,
    };
  }
  let expected: string;
  try {
    expected = acceleratorLockFingerprint(directory, prepared);
  } catch (cause) {
    return {
      status: "warning",
      detail: `the ${prepared} environment could not be checked: ${cause instanceof Error ? cause.message : String(cause)}`,
      rebuilt: false,
      prepared,
    };
  }
  let recorded = "";
  try {
    const payload = JSON.parse(fs.readFileSync(acceleratorRecordPath(), "utf8")) as {
      lock_fingerprint?: unknown;
    };
    recorded = String(payload.lock_fingerprint ?? "");
  } catch {
    recorded = "";
  }
  if (recorded === expected) {
    return {
      status: "skipped",
      detail: `the ${prepared} environment is unchanged since the last build`,
      rebuilt: false,
      prepared,
    };
  }
  if (skipAccelerator) {
    return {
      status: "warning",
      detail: `the ${prepared} environment was resolved from an older lockfile and keeps serving as it is; rebuild it with \`code-indexing-mcp configure --accelerator ${prepared}\``,
      rebuilt: false,
      prepared,
    };
  }
  try {
    const plan = await configureAccelerator(directory, prepared);
    return {
      status: plan.honored ? "ok" : "warning",
      detail: `${plan.accelerator} (${plan.reason})`,
      rebuilt: true,
      prepared: ACCELERATOR_PROBE_TARGETS.has(plan.accelerator) ? plan.accelerator : null,
    };
  } catch (cause) {
    return {
      status: "warning",
      detail: `the ${prepared} environment could not be rebuilt: ${cause instanceof Error ? cause.message : String(cause)}`,
      rebuilt: false,
      prepared,
    };
  }
}

async function finalizeMain(
  directory: string,
  options: { previousSha: string | null; skipAccelerator: boolean; runCommand: RunCommand },
): Promise<number> {
  const head = spawnSync("git", ["rev-parse", "HEAD"], {
    cwd: directory,
    encoding: "utf8",
  }).stdout.trim();
  const accelerator = await reconcileAccelerator(directory, options.skipAccelerator);
  printStatus("accelerator", accelerator.status, accelerator.detail);
  const expectedCommand = serverExecutable(directory);
  const configuredSlugs = [...loadPrefill().configuredSlugs];
  const owned: Record<string, boolean> = {};
  for (const slug of configuredSlugs) {
    const entry = readServerEntry(slug);
    const command = entry === null ? null : commandFromEntry(slug, entry);
    owned[slug] = command !== null && command === expectedCommand;
  }
  const skillGroups = new Map<string, string[]>();
  for (const slug of configuredSlugs) {
    const skills = skillDirectory(slug);
    if (skills === null) continue;
    let groupKey = skills;
    try {
      groupKey = fs.realpathSync(skills);
    } catch {
      groupKey = path.resolve(skills);
    }
    const group = skillGroups.get(groupKey) ?? [];
    group.push(slug);
    skillGroups.set(groupKey, group);
  }
  const installSlugs: string[] = [];
  for (const [skills, group] of skillGroups) {
    const ours = group.filter((slug) => owned[slug] === true);
    const others = group.filter((slug) => owned[slug] !== true);
    if (ours.length > 0 && others.length > 0) {
      for (const slug of ours) {
        printStatus(
          "skills",
          "warn",
          `${slug}: skipped: ${skills} is shared with ${others.join(", ")} configured for another installation`,
        );
      }
    } else {
      installSlugs.push(...ours);
    }
  }
  const ownedSlugs = configuredSlugs.filter((slug) => owned[slug] === true);
  const configured = ownedSlugs.map((slug) => [slug, configurationPath(slug)] as [string, string]);
  for (const [slug, message] of installSkills(installSlugs, directory)) {
    printStatus("skills", message.startsWith("skipped:") ? "warn" : "ok", `${slug}: ${message}`);
  }
  const checks = runUpdateChecks(directory, configured, {
    acceleratorWasPrepared: accelerator.prepared !== null,
  });
  for (const check of checks) {
    (check.ok ? process.stdout : process.stderr).write(`${formatCheck(check)}\n`);
  }
  if (options.previousSha !== null && options.previousSha !== head) {
    process.stdout.write(`Updated ${options.previousSha.slice(0, 7)} -> ${head.slice(0, 7)}\n`);
    try {
      const lines = git(
        options.runCommand,
        directory,
        "log",
        "--oneline",
        "--no-decorate",
        `${options.previousSha}..HEAD`,
      ).split(/\r?\n/);
      for (const line of lines.slice(0, LOG_LINE_LIMIT)) process.stdout.write(`  ${line}\n`);
      if (lines.length > LOG_LINE_LIMIT) {
        process.stdout.write(`  ... and ${lines.length - LOG_LINE_LIMIT} more commits\n`);
      }
    } catch {
      // Summary is best-effort.
    }
  } else {
    process.stdout.write(
      `Already at ${head.slice(0, 7)}; the environment was re-synced and re-checked.\n`,
    );
  }
  process.stdout.write(`Accelerator: ${accelerator.detail}\n`);
  process.stdout.write("Restart your MCP clients to load the updated server.\n");
  return checks.some((check) => check.status === "fail" && check.name === "server executable")
    ? 1
    : 0;
}

function checkMain(directory: string, runCommand: RunCommand): number {
  if (!fs.existsSync(path.join(directory, ".git"))) {
    return error(`${directory} is not a git checkout; there is no update to check for`);
  }
  let origin: string;
  try {
    origin = git(runCommand, directory, "remote", "get-url", "origin");
  } catch (cause) {
    return error(
      `${directory} has no origin remote to check against: ${cause instanceof Error ? cause.message : String(cause)}`,
    );
  }
  const expected = process.env.CODE_INDEXING_MCP_REPO_URL ?? DEFAULT_REPOSITORY_URL;
  if (canonicalRepositoryUrl(origin) !== canonicalRepositoryUrl(expected)) {
    return error(`the checkout at ${directory} tracks ${origin}, not ${expected}`);
  }
  const result = spawnSync("git", ["ls-remote", "origin", "refs/heads/main"], {
    cwd: directory,
    encoding: "utf8",
    timeout: LS_REMOTE_TIMEOUT_SECONDS * 1000,
  });
  if (result.error !== undefined || result.status !== 0) {
    return error(
      `could not reach origin to check for updates: ${result.error?.message ?? result.stderr}`,
    );
  }
  const lines = result.stdout.split(/\r?\n/).filter((line) => line.trim() !== "");
  if (lines.length === 0) return error("origin has no refs/heads/main to compare against");
  const remoteSha = (lines[0] ?? "").split(/\s+/)[0] ?? "";
  const localSha = spawnSync("git", ["rev-parse", "HEAD"], {
    cwd: directory,
    encoding: "utf8",
  }).stdout.trim();
  process.stdout.write(
    `${JSON.stringify(
      {
        install_dir: directory,
        local_sha: localSha,
        remote_sha: remoteSha,
        update_available: localSha !== remoteSha,
      },
      null,
      2,
    )}\n`,
  );
  return localSha !== remoteSha ? CHECK_UPDATE_AVAILABLE_EXIT : 0;
}
